nano/node/confirmation_height_processor.cpp

```diff
@@ -0,0 +1,371 @@
+			}
+			assert (pending_writes.empty ());
+		}
+		receive_source_pairs_lk.unlock ();
+		// Exit early when the processor has been stopped, otherwise this function may take a
+		// while (and hence keep the process running) if updating a long chain.
+		if (stopped)
+		{
+			break;
+		}
+		receive_source_pairs_lk.lock ();
+	} while (!receive_source_pairs.empty ());
+}
+
+// Returns true if there was an error in finding one of the blocks to write the confirmation height for.
+bool nano::confirmation_height_processor::write_pending (std::deque<conf_height_details> & all_pending, int64_t total_pending_write_block_count_a)
+{
+	nano::account_info account_info;
+	auto total_pending_write_block_count (total_pending_write_block_count_a);
+
+	// Write in batches
+	while (total_pending_write_block_count > 0)
+	{
+		uint64_t num_block_writes = 0;
+		auto transaction (store.tx_begin_write ());
+		while (!all_pending.empty ())
+		{
+			const auto & pending = all_pending.front ();
+			auto error = store.account_get (transaction, pending.account, account_info);
+			release_assert (!error);
+			if (pending.height > account_info.confirmation_height)
+			{
+#ifndef NDEBUG
+				// Do more thorough checking in Debug mode
+				nano::block_sideband sideband;
+				auto block = store.block_get (transaction, pending.hash, &sideband);
+				assert (block != nullptr);
+				assert (sideband.height == pending.height);
+				// Check that the block still exists as there may have been changes outside this processor.
+				if (!block)
+				{
+					logger.always_log ("Failed to write confirmation height for: ", pending.hash.to_string ());
+					ledger.stats.inc (nano::stat::type::confirmation_height, nano::stat::detail::invalid_block);
+					return true;
+				}
+#endif
+				ledger.stats.add (nano::stat::type::confirmation_height, nano::stat::detail::blocks_confirmed, nano::stat::dir::in, pending.height - account_info.confirmation_height);
+				assert (pending.num_blocks_confirmed == pending.height - account_info.confirmation_height);
+				account_info.confirmation_height = pending.height;
+				store.account_put (transaction, pending.account, account_info);
+			}
+			total_pending_write_block_count -= pending.num_blocks_confirmed;
+			num_block_writes += pending.num_blocks_confirmed;
+			all_pending.erase (all_pending.begin ());
+
+			if (num_block_writes >= batch_write_size)
+			{
+				// Commit changes periodically to reduce time holding write locks for long chains
+				break;
+			}
+		}
+	}
+	return false;
+}
+
+// This function assumes receive_source_pairs_lk is not already locked
+void nano::confirmation_height_processor::collect_unconfirmed_receive_and_sources_for_account (uint64_t block_height, uint64_t confirmation_height, nano::block_hash const & current, const nano::block_hash & genesis_hash, std::vector<receive_source_pair> & receive_source_pairs, nano::account const & account, nano::transaction & transaction, std::unique_lock<std::mutex> & receive_source_pairs_lk)
```

Comment:
I guess genesis_hash is not required anymore here & in other parts of confirmation_height_processor
Also isn't receive_source_pairs vector available for all confirmation_height_processor functions, so you shoudn't pass it, right?

Reply:
good points! I made it a member so that I could access it in the memory stats RPC, have removed unnecessary parameters from the function.